### PR TITLE
Updating Readme file

### PR DESCRIPTION
**Pull request by sugudsl**
Readme file updated

### CODE REVIEW

File: thinklmi-kernel/think-lmi.c [hunk 1]
```diff
@@ -893,6 +893,8 @@ static void think_lmi_analyze(struct think_lmi *think)
 		int num = 0;
 		char *p;
 
+		printk("iteration %d\n",i);
```
Comment (mrhpearson) [inline thinklmi-kernel/think-lmi.c:896]: Please remove unneeded debug.

File: thinklmi-kernel/think-lmi.c [hunk 2]
```diff
@@ -915,9 +917,15 @@ static void think_lmi_analyze(struct think_lmi *think)
 		if (p)
 			*p = '\0';
 		think->settings[i] = item; /* Cache setting name */
+                printk("%d  %s\n", i, think->settings[i]);
```
Comment (mrhpearson) [inline thinklmi-kernel/think-lmi.c:920]: remove this

File: thinklmi-kernel/think-lmi.c [hunk 3]
```diff
@@ -915,9 +917,15 @@ static void think_lmi_analyze(struct think_lmi *think)
...
+
 		think->settings_count++;
+
+		printk("setting count is %d\n",think->settings_count);
```
Comment (mrhpearson) [inline thinklmi-kernel/think-lmi.c:924]: ditto

File: thinklmi-kernel/think-lmi.c [hunk 4]
```diff
@@ -915,9 +917,15 @@ static void think_lmi_analyze(struct think_lmi *think)
...
 	}
 
+	printk("final count %d\n", think->settings_count);
+
```
Comment (mrhpearson) [inline thinklmi-kernel/think-lmi.c:928]: ditto

File: thinklmi-user/thinklmi.c [hunk 5]
```diff
@@ -39,7 +39,7 @@ void get_settings_all(int fd)
         perror("query_apps ioctl get");
     } else {
 	printf("Total settings: %d\n", settings_count);
-	for(i=0; i <= settings_count; i++)
+	for(i=0; i <= 0xFF; i++)
```
Comment (mrhpearson) [inline thinklmi-user/thinklmi.c:42]: I really don't like this - why is this being done?

**Review by mrhpearson — CHANGES_REQUESTED** [review]
The biggest issue is the change to scan the settings array to a hard-coded 0xFF. This seems very wrong to me.
The other review comments are fairly minor.
Let me know if any questions

File: thinklmi-user/README.md [hunk 6]
```diff
@@ -63,6 +63,52 @@ eg: ./thinklmi -p hello newpd pap ascii us
 Change the supervisor password or power on password using this command.
 Do a reboot for the new password to take effect.
 
+## Debug
+./thinklmi -d [BIOS Setting] [option]
+
+This configures a the BIOS setting in debug mode. Use with care.
```
[the remark this answers is not in the document]
Comment (sugudsl) [reply thinklmi-user/README.md:69]: Done

File: thinklmi-user/README.md [hunk 7]
```diff
@@ -63,6 +63,52 @@ eg: ./thinklmi -p hello newpd pap ascii us
...
+
+eg: ./thinklmi -d FnSticky Enable
+
+
+## Change Password using LMIOPCODE
+./thinklmi -w [Admin Password] [Password Type] [Old Password] [New Pasword]
+
+This command changes the password in thinkstation systems, with Admin password
```
[the remark this answers is not in the document]
Comment (sugudsl) [reply thinklmi-user/README.md:77]: The password type can be any BIOS supported password. Not just pap and pop. So changing it like this.
The password type can be any BIOS supported password like pop, pap, mhdp, uhdp etc.

File: thinklmi-user/README.md [hunk 8]
```diff
@@ -63,6 +63,52 @@ eg: ./thinklmi -p hello newpd pap ascii us
...
+./thinklmi -w [Password Type] [Old Password] [New Password]
+
+This command changes the password in thinkstation systems, without an Admin password
+
+eg: ./thinklmi -w pop oldpass newpass
+
+## TPMTYPE Change
+./thinklmi -t [TPM Type]
+
+This command changes the TPM Type in thinkstation systems
```
[the remark this answers is not in the document]
Comment (sugudsl) [reply thinklmi-user/README.md:91]: Done.

File: thinklmi-user/README.md [hunk 9]
```diff
@@ -63,6 +63,52 @@ eg: ./thinklmi -p hello newpd pap ascii us
...
+
+./thinklmi -t tpm12
+
+
+## Load Default Settings
+./thinklmi -l
+
+This command loads the Default BIOS Settings
+
```
[the remark this answers is not in the document]
Comment (sugudsl) [reply thinklmi-user/README.md:100]: there was an extra space after this. Did you mean that?. It seems OK now

File: thinklmi-kernel/think-lmi.c (hunk 1 above)
Comment (sugudsl) [reply thinklmi-kernel/think-lmi.c:896]: done

File: thinklmi-kernel/think-lmi.c (hunk 2 above)
Comment (sugudsl) [reply thinklmi-kernel/think-lmi.c:920]: done

File: thinklmi-kernel/think-lmi.c (hunk 3 above)
Comment (sugudsl) [reply thinklmi-kernel/think-lmi.c:924]: done

File: thinklmi-kernel/think-lmi.c (hunk 4 above)
Comment (sugudsl) [reply thinklmi-kernel/think-lmi.c:928]: done

File: thinklmi-user/thinklmi.c (hunk 5 above)
Comment (sugudsl) [reply thinklmi-user/thinklmi.c:42]: replaced with TLMI_MAX_SETTINGS
The settings count was missing some of the settings at the end.
So we have to parse through the entire settings and display the valid ones

Comment (mrhpearson) [reply thinklmi-user/thinklmi.c]: Gotcha - this is better.

Can you please change higher in the file where settings is declared to also use this define
	unsigned char *settings[256];
To keep everything consistent :)

**Comment by mrhpearson** [comment]
Just curious - why did you remove the discard settings and the nopap options? Are there issues with the implementation?

Thanks for all the fixes - almost there

Mark

**Comment by sugudsl** [comment]
> Just curious - why did you remove the discard settings and the nopap options? Are there issues with the implementation?
> 
> Thanks for all the fixes - almost there
> 
> Mark

The BIOS team wants the user to set an Admin Password and only then change other passwords.

Regarding the Discard settings, they say, though it is in the specification, the BIOS does not support it

File: thinklmi-user/thinklmi.c (hunk 5 above)
Comment (sugudsl) [reply thinklmi-user/thinklmi.c]: Done

**Pull request merged**